Fix for #84

alexjgriffith opened the pull request:
This PR lets us use `mastodon-mode` commands to update timeline like buffers, such as `*mastodon-notifications*` and `*mastodon-profile-user*`, and give the current mastodon buffers more user friendly names in the future (`*mastodon-local*` vs `*mastodon-public?local=true*`.

# Current Version

We now create a mastodon buffer using `mastodon-tl--init`. We pass it  a `buffer-name` an `endpoint` and an `update-function.
```
 (mastodon-tl--init
   "home" "timelines/home" 'mastodon-tl--timeline))
```

`mastodon-tl--init` overwrites the global `mastodon-buffer-spec` with a local buffer variable. If the buffer is not set localy its value will be `nil`
``` elsip
(defvar mastodon-buffer-spec nil)
```

`mastodon-buffer-spec` is a local buffer plist containing the "parsed" `buffer-name`, `endpoint` and `update-function`. There are getters that can be used in an mastodon-buffer to access these elements. 

``` elisp
(make-local-variable 'mastodon-buffer-spec)
(setq mastodon-buffer-spec
         (buffer-name "home" endpoint "timelines/home"
                                update-function mastodon-tl--timeline)
```

# Previous version
For simplicity this method still relies on the buffer naming convention. This is a liability and probably should be changed in the future to our own buffer local variable.

The parsed buffer name is matched against a list of plists.  A buffer specific endpoint is then passed to `mastodon-tl--more-json` and a buffer specific update function is used in `mastodon-tl--more`, rather than calling `mastodon-tl--timelines` directly. This list of plists is currently contained within `mastodon-tl--get-endpoint` but could be moved to a global variable so each buffer type could append their own endpoints and update function.

## Example mastodon-buffer-endpoints
``` elisp
          '((buffer-name "home" endpoint "timelines/home"
                         update-function mastodon-tl--timeline)
            (buffer-name "local" endpoint "timelines/public?local=true"
                         update-function 'mastodon-tl--timeline)
            (buffer-name "federated" endpoint "timelines/public"
                         update-function 'mastodon-tl--timeline)
            (buffer-name "public?local=true"
                         endpoint "timelines/public?local=true"
                         update-function 'mastodon-tl--timeline)
            (buffer-name "public"
                         endpoint "timelines/public"
                         update-function 'mastodon-tl--timeline)
            (buffer-name "notifications"
                         endpoint "notifications"
                         update-function
                         mastodon-notifications--notifications))
```

side note, renaming the `*mastodon-public?local=true*` buffer to `*mastodon-local*` works like a charm!

alexjgriffith committed:
Fix for #84

changed more-json in mastodon-tl-tests to recieve an endpont

more evocative names and fixed whitespace

replaced remaining end eith endpoint-plist

added basic unit tests for endpoint selection

replaced equalp with equal in tests

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -212,22 +212,72 @@ also render the html"
+                  ((stringp (plist-get endpoint-plist 'endpoint))
+                   endpoint-plist)
+                  ((functionp (plist-get endpoint-plist 'endpoint))
+                   (plist-put end 'endpoint
```
This still says `end` instead of `endpoint-plist`.

This strongly suggests that it is not covered by any of your tests.   ;-)

alexjgriffith replied:
I.e. perhaps I should write tests :p

Edit: Fixed for now. Will add unit tests tomorrow

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -216,22 +216,72 @@ also render the html"
+                  (t (error
+                      (format "%s not a string or function"
+                              (plist-get endpoint-plist 'endpoint))))))))
+         (endpoint (remove-if 'null
```
`remove-if` seems to be part of cl.el and the tests seem to fail because of that.

alexjgriffith replied:
I will replace remove-if, but the builds only stopped passing once i added my own unit tests.

Commit:
Putting buffer name and update function in buffer local var

removed tests for previous version

documented mastodon-buffer-spec getters

moved mastodon-buffer-spec to mastodon.el

combined mastodon-tl--get and mastodon-tl--init, along with other cleanups

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -223,22 +221,50 @@ also render the html"
+(defun mastodon-tl--buffer-name ()
+  "Get the BUFFER-NAME stored in `mastodon-buffer-spec'"
+  (mastodon-tl--get-buffer-property 'buffer-name
+				    (or buffer (current-buffer))))
```
There is no variable `buffer` here.
Missing the `&optional buffer` above, I guess.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -223,22 +221,50 @@ also render the html"
+(defun mastodon-tl--get-update-function (&optional buffer)
+  "Get the UPDATE-FUNCTION stored in `mastodon-buffer-spec'"
+  (mastodon-tl--get-buffer-property 'update-function
+				    (or buffer (current-buffer))))
```
`mastodon-tl--get-buffer-property` already handles a null buffer so no need to to the `or` here as well.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -281,39 +307,54 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
+  "Initialize BUFFER-NAME with timeline targeted by ENDPOINT.
 
-(defun mastodon-tl--get (timeline)
+UPDATE-FUNCTION is used to recive more toots."
```
"receive"

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -281,39 +307,54 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
+                          endpoint ,endpoint update-function
+                          ,update-function)))))
+
+(defun mastodon-tl--get (buffer-name endpoint update-function)
```
That name really needs improving.

alexjgriffith replied:
I think I am going to envelop `mastodon-tl--get` into `mastodon-tl--init` since it has no purpose on its own.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -281,39 +307,54 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
-         (json (mastodon-tl--more-json tl id)))
-    (when json
-      (with-current-buffer (current-buffer)
+  (with-current-buffer (current-buffer)
```
?!?  
`(current-buffer)` *is* already the current buffer.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -281,39 +307,54 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
-          (mastodon-tl--timeline json)
+          (funcall update-function json)
           (goto-char point-before)
           (mastodon-tl--goto-next-toot))))))
```
If the `update-function` won't provide us with more data, then this will be an infinite loop, right?  (`mastodon-tl--goto-next-toot` will notice that there is no more data and recurse.  Or will mastodon fail and not return any json in that case?)

Note: this is not a new issue but may become one more urgently as we use this function for other types of timelines that *can* easily run out of old data.

alexjgriffith replied:
Good catch, I will remove the goto-next toot line at the end and leave the pointer in its previous location.

hdurer commented on lisp/mastodon-tl.el:
```diff
@@ -281,39 +307,54 @@ Move forward (down) the timeline unless BACKWARD is non-nil."
-         (json (mastodon-tl--updated-json tl id)))
-    (when json
-      (with-current-buffer (current-buffer)
+  (with-current-buffer (current-buffer)
```
Again.

hdurer committed:
Remove most byte-compile warnings.

We do this by
- moving vars into the files where they are (mostly) used
- "declaring" vars used elsewhere with the (defvar <var-name>) pattern,
- declaring functions defined in others functions rather than loading the file via require.

hdurer committed:
Replace uses of replace-regexp with search-forward and replace-match.

jdenen reviewed (commented):
When I try to open the home timeline, I get the following error:

```
Debugger entered--Lisp error: (void-function mastodon-media--get-media-link-rendering)
  (mastodon-media--get-media-link-rendering preview-url)
 ```

From the `mastodon-tl--media` function.

Commit:
replaced declare-function with autoload
